File: memory/native/state/memory_state.c
#include "memory_state.h"
#include <buffers/memory_io_buffers.h>
#include <buffers/memory_static_buffers.h>
#include <memory/memory.h>
#include <system/system.h>

int32_t memory_state_create(struct memory_state* state, struct memory_configuration* configuration)
{
    state->memory_instance = memory_module_new(sizeof(struct memory));
    if (state->memory_instance == NULL)
    {
        return -ENOMEM;
    }
    if (memory_create(state->memory_instance, configuration->quota_size, configuration->preallocation_size, configuration->slab_size))
    {
        return -ENOMEM;
    }

    state->static_buffers = memory_module_new(sizeof(struct memory_static_buffers));
    if (state->static_buffers == NULL)
    {
        return -ENOMEM;
    }

    state->static_buffers = memory_static_buffers_create(configuration->static_buffers_capacity, configuration->static_buffer_size);
    if (state->static_buffers != NULL)
    {
        return -ENOMEM;
    }

    state->io_buffers = memory_module_new(sizeof(struct memory_io_buffers));
    if (!state->io_buffers)
    {
        return -ENOMEM;
    }

    state->io_buffers = memory_io_buffers_create(state->memory_instance);
    if (state->io_buffers != NULL)
    {
        return -ENOMEM;
    }

    return 0;
}

void memory_state_destroy(struct memory_state* state)
{
    memory_static_buffers_destroy(state->static_buffers);
    memory_io_buffers_destroy(state->io_buffers);
    memory_destroy(state->memory_instance);
    free(state->static_buffers);
    free(state->io_buffers);
    free(state->memory_instance);
}

File: memory/dart/lib/bindings/state/memory_state.dart
import 'dart:ffi';

import '../include/memory.dart';
import '../include/memory_configuration.dart';
import '../include/memory_io_buffers.dart';
import '../include/memory_static_buffers.dart';

final class memory_state extends Struct {
  external Pointer<memory_static_buffers> static_buffers;
  external Pointer<memory_io_buffers> io_buffers;
  external Pointer<memory> memory_instance;
}

@Native<Pointer<memory_state> Function()>(symbol: 'memory_state_construct', assetId: 'memory-bindings', isLeaf: true)
external Pointer<memory_state> memory_state_construct();

@Native<Void Function(Pointer<memory_state>)>(symbol: 'memory_state_destruct', assetId: 'memory-bindings', isLeaf: true)
external void memory_state_destruct(Pointer<memory_state> memory);

@Native<Int32 Function(Pointer<memory_state>, Pointer<memory_configuration>)>(symbol: 'memory_state_create', assetId: 'memory-bindings', isLeaf: true)
external int memory_state_create(Pointer<memory_state> memory, Pointer<memory_configuration> configuration);

@Native<Void Function(Pointer<memory_state>)>(symbol: 'memory_state_destroy', assetId: 'memory-bindings', isLeaf: true)
external void memory_state_destroy(Pointer<memory_state> memory);
